### PR TITLE
Fix/error handling

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: https://github.com/storybookjs/storybook/issues/10577 https://github.com/storybookjs/storybook/issues/10664

when there's an error in one of the stories initially, the preview & sidebar will go in a perpetual loading-state & error doesn't show up in the preview

## What I did

- made sure that if an error occurs storybook preview renders an error.
- Re-use the UI from refs to display the error in sidebar
- Add the error that occurred to setStories event

### CODE REVIEW

**Comment by ndelangen** [comment]
This is what an error now looks like when loading stories fails:
Screenshot 2020-05-06 at 09 22 38
Screenshot 2020-05-06 at 09 23 43

This is what it looks like when a story-file is loaded successfully, but the story throws an error being rendered:
Screenshot 2020-05-06 at 09 21 39

File: lib/api/src/modules/stories.ts [hunk 1]
```diff
@@ -382,6 +385,38 @@ export const init: ModuleFn = ({
       (storiesHash[id] as Story).args = args;
       store.setState({ storiesHash });
     });
+
+    fullAPI.on(CURRENT_STORY_WAS_SET, function handleSetStory({ storyId, viewMode }) {
```
Comment (tmeasday) [inline lib/api/src/modules/stories.ts:389]: We already have events to indicate that configuring is done:
  A) In 6.0, the `SET_STORIES` event, which is guaranteed to run once when the configure process has run. I'm ok with adding `error` to it though if that makes sense.

  B) In 5.x the `STORIES_CONFIGURED` event, I'm not sure what runs when there is an error however.

File: lib/client-api/src/config_api.ts [hunk 2]
```diff
@@ -27,6 +27,7 @@ export default class ConfigApi {
       this._storyStore.finishConfiguring();
     } catch (err) {
       this._storyStore.setError(err);
+      this._storyStore.failConfiguring(err);
```
Comment (tmeasday) [inline lib/client-api/src/config_api.ts:30]: Why can't we call `finishConfiguring()`? The store knows there is an error itself.

File: lib/client-api/src/story_store.test.ts [hunk 3]
```diff
@@ -793,16 +793,18 @@ describe('preview.story_store', () => {
     });
   });
 
-  describe('RENDER_CURRENT_STORY', () => {
+  describe('STORY_ERRORED', () => {
     it('is emitted when setError is called', () => {
       const onRenderCurrentStory = jest.fn();
-      channel.on(Events.RENDER_CURRENT_STORY, onRenderCurrentStory);
+      channel.on(Events.STORY_ERRORED, onRenderCurrentStory);
```
Comment (tmeasday) [inline lib/client-api/src/story_store.test.ts:799]: This is semantically wrong. No story errored, there was a problem configuring the store.

We use `RENDER_CURRENT_STORY` here to tell the `StoryRenderer` to figure out what to render to the screen.

File: lib/client-api/src/story_store.ts [hunk 4]
```diff
@@ -152,10 +152,25 @@ export default class StoryStore {
     this._configuring = true;
   }
 
+  failConfiguring(error: Error) {
```
Comment (tmeasday) [inline lib/client-api/src/story_store.ts:155]: We can just do this in `finishConfiguring` if `this._error` is set.

File: lib/client-api/src/story_store.ts [hunk 5]
```diff
@@ -481,7 +496,10 @@ export default class StoryStore {
 
   setError = (err: ErrorLike) => {
     this._error = err;
-    if (this._channel) this._channel.emit(Events.RENDER_CURRENT_STORY);
+
+    if (this._channel) {
+      this._channel.emit(Events.STORY_ERRORED, { message: err.message, stack: err.stack });
```
Comment (tmeasday) [inline lib/client-api/src/story_store.ts:501]: We shouldn't emit anything at this point I don't think.

**Review by tmeasday — CHANGES_REQUESTED** [review]
I'm OK with the idea of including `errors` in `SetStoriesPayload` (if we need to know about it on the manager side).

But I don't like the changed event behaviour. Let's avoid incremental changes to events like this, that's how we ended up with the complex mess of incoherent events we had before. 

Let's be careful about what events are emitted when and document things. I tried to do that already although it could be more comphrehensive I think: https://github.com/storybookjs/storybook/blob/a9fc57fe490e5befcb4bdd4f77f6897d283d8263/lib/core/README.md#events-on-startup

File: lib/api/src/lib/stories.ts [hunk 6]
```diff
@@ -107,6 +107,7 @@ export interface StoriesRaw {
 export interface SetStoriesPayload {
   v?: number;
   stories: StoriesRaw;
+  error?: Error;
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/api/src/lib/stories.ts:110]: ow true!

File: lib/client-api/src/story_store.ts (hunk 4 above)
Comment (ndelangen) [reply lib/client-api/src/story_store.ts:155]: ok, great!

**Comment by yannbf** [comment]
I don't have 100% context so can't help that much in code, but I ran the official storybook in a few scenarios:
1 - with a wrong glob in main.js:
```
stories: ['../../lib/ui/src/**/Wrong.stories.(js|tsx|mdx)']
```

2 - with a correct glob but everything commented out in the story:
```
stories: ['../../lib/ui/src/**/Sidebar.stories.(js|tsx|mdx)']
```

3 - with a correct glob, exporting a single story but not default exporting anything

Scenarios 1, 2 and 3 leave me in this screen:
image

4 - with a correct glob + the file only has a `throw new Error('error!!!');`
5 - with a correct glob + the file has a story that throws an error

Scenarios 4 and 5 work as you mentioned in the PR.

And a small observation, that you can ignore if you want, the component `/components/sidebar/RefBlocks -> ErrorBlock` has the following text:
> Ow now! something went wrong loading this storybook

Shouldn't it be 

> Oh no! Something went wrong loading this storybook.

File: lib/ui/src/components/sidebar/Sidebar.tsx [hunk 7]
```diff
@@ -75,10 +77,22 @@ const useSearchResults = (refsList: [string, RefType][], filter: string) => {
   return { total: refsTotal || 0, list: refsLengths };
 };
 
-const useCombination = (stories: StoriesHash, refs: Refs) => {
+const useCombination = (
+  stories: StoriesHash,
+  ready: boolean,
+  error: Error | undefined,
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/ui/src/components/sidebar/Sidebar.tsx:83]: This is not an interface?

[the remark this answers is not in the document]
Comment (yannbf) [reply lib/ui/src/components/sidebar/Sidebar.tsx:83]: Indeed, but the optional notation can also be used in parameters. 
However, in this case it won't actually work unless `error` is the last parameter. Optionals have to be placed at the end of the parameters

[the remark this answers is not in the document]
Comment (yannbf) [reply lib/ui/src/components/sidebar/Sidebar.tsx:83]: it was just a suggestion, feel free to ignore it!

**Comment by ndelangen** [comment]
I'll investigate the scenarios:

- **all stories load fine, but they don't export/define any stories**
  should show a message in the sidebar
- **story-files are missing (glob matches nothing)**
  should show a warning in terminal?

File: lib/ui/src/components/preview/preview.tsx [hunk 8]
```diff
@@ -57,7 +61,8 @@ const createCanvas = (id: string, baseUrl = 'iframe.html', withLoader = true): A
           ]);
 
           const isLoading = !!(
-            (storyId && !story) ||
+            (!story && !storiesFailed) ||
+            (!story && storiesConfigured === false) ||
             (story && story.refId && refs[refId] && !refs[refId].ready)
```
Comment (shilman) [inline lib/ui/src/components/preview/preview.tsx:66]: simpoler: `(refId && refs[refId] && ! refs[refId].ready)`

File: lib/client-api/src/story_store.ts [hunk 9]
```diff
@@ -481,7 +484,14 @@ export default class StoryStore {
...
     this._error = err;
-    if (this._channel) this._channel.emit(Events.RENDER_CURRENT_STORY);
+
+    if (this._channel) {
+      // If the selection is set while configuration is in process, we are guaranteed
+      // we'll emit RENDER_CURRENT_STORY at the end of the process, so we shouldn't do it now.
+      if (!this._configuring) {
+        this._channel.emit(Events.RENDER_CURRENT_STORY);
+      }
+    }
```
Comment (tmeasday) [inline lib/client-api/src/story_store.ts:494]: I don't think it is allowed that you set error outside of configuration. In fact we should probably throw if you do that.

Comment (ndelangen) [reply lib/client-api/src/story_store.ts:494]: AFAIK the setError is also called when a story throws?

Comment (tmeasday) [reply lib/client-api/src/story_store.ts:494]: Who by?

**Comment by ndelangen** [comment]
Here's what happens if a ref/frame results in a successful SET_STORIES, but the contents are empty:

Screenshot 2020-05-07 at 17 31 45

**Comment by tmeasday** [comment]
Maybe we can make a better UI for the no stories case? WDYT?

**Comment by ndelangen** [comment]
Screenshot 2020-05-08 at 20 35 56

new UI

**Comment by tmeasday** [comment]
Looking good @ndelangen. I was actually thinking something similar in the preview pane. Wdyt?

**Comment by tmeasday** [comment]
@domyen want to weigh in?

**Pull request merged**